### ISSUE TITLE
Race-condition and early destruction in non-blocking HIP queues

### ISSUE THREAD

**Issue by tdd11235813** [alpaka-group/alpaka#805]
A deadlock can happen in the HIP(NVCC) non-blocking queue, where a spawned thread waits for the queue mutex before task-execution, while another one waits for the execution of its stream callback, which occasionally can come after the first callback and yields the deadlock.
This does not happen for HIP(HCC) because the HCC stream is not blocked by callbacks' execution.
                                                                                                                                        
It also could happen, that the HIP non-blocking queue can be destroyed (going out-of-scope) while tasks are still executing. The HIP queue workaround introduced the capture of the queue object, so I have to ensure, it exists until the last task finished.

I already have a first solution in the pipe, but it has side effects, so I need some time to look for a better solution :)
